spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/repository/query/PartTreeDatastoreQuery.java

```diff
@@ -0,0 +1,212 @@
+				.getPersistentEntity(this.entityType);
+
+		if (this.tree.isDelete()) {
+			throw new UnsupportedOperationException(
+					"Delete queries are not supported in Cloud Datastore: "
+							+ this.queryMethod.getName());
+		}
+		else if (this.tree.isDistinct()) {
+			throw new UnsupportedOperationException(
+					"Cloud Datastore structured queries do not support the Distinct keyword.");
+		}
+
+		List parts = this.tree.getParts().get().collect(Collectors.toList());
+		if (parts.size() > 0) {
+			if (parts.get(0) instanceof OrPart) {
+				throw new DatastoreDataException(
+						"Cloud Datastore only supports multiple filters combined with AND.");
+			}
+			this.filterParts = parts;
+		}
+		else {
+			this.filterParts = Collections.emptyList();
+		}
+	}
+
+	@Override
+	public Object execute(Object[] parameters) {
+		List<T> results = executeRawResult(parameters);
+		if (this.tree.isCountProjection()) {
+			return results.size();
+		}
+		else if (this.tree.isExistsProjection()) {
+			return !results.isEmpty();
+		}
+		else {
+			return applyProjection(results);
+		}
+	}
+
+	@Override
+	List<T> executeRawResult(Object[] parameters) {
+		Iterable<T> found = this.datastoreOperations.query(getQuery(parameters),
+				this.entityType);
+		return found == null ? Collections.emptyList()
+				: StreamSupport.stream(found.spliterator(), false)
+						.collect(Collectors.toList());
+	}
+
+	private StructuredQuery<Entity> getQuery(Object[] parameters) {
+		Builder<Entity> builder = StructuredQuery.newEntityQueryBuilder();
+
+		builder.setKind(this.datastorePersistentEntity.kindName());
+
+		if (this.tree.hasPredicate()) {
+			builder.setFilter(getFilter(parameters));
+		}
+
+		if (!this.tree.getSort().isUnsorted()) {
+			applySort(builder);
+		}
+
+		if (this.tree.isLimiting()) {
+			builder.setLimit(this.tree.getMaxResults());
+		}
+
+		return builder.build();
+	}
+
+	private void applySort(Builder builder) {
+		OrderBy[] orders = this.tree.getSort().get().map(sort -> {
+			String fieldName = ((DatastorePersistentProperty) this.datastorePersistentEntity
+					.getPersistentProperty(sort.getProperty())).getFieldName();
+			return sort.isAscending() ? OrderBy.asc(fieldName) : OrderBy.desc(fieldName);
+		}).toArray(OrderBy[]::new);
+		if (orders.length > 1) {
+			builder.setOrderBy(orders[0], Arrays.copyOfRange(orders, 1, orders.length));
+		}
+		else {
+			builder.setOrderBy(orders[0]);
+		}
+	}
+
+	private Filter getFilter(Object[] parameters) {
+		Iterator it = Arrays.asList(parameters).iterator();
+		Filter[] filters = this.filterParts.stream().map(part -> {
+			Filter filter;
+			String fieldName = ((DatastorePersistentProperty) this.datastorePersistentEntity
+					.getPersistentProperty(part.getProperty().getSegment()))
+							.getFieldName();
+			try {
+				switch (part.getType()) {
+				case IS_NULL:
+					filter = PropertyFilter.isNull(fieldName);
+					break;
+				case IS_EMPTY:
+					filter = PropertyFilter.isNull(fieldName);
```

Comment:
This should probably have a different meaning than `IS_NULL`. Compare with other implementations like Cassandra.

Reply:
I checked cassandra, jpa, and mongodb. None of those 3 seem to support IS_EMPTY. I will leave it out too.

Reply:
https://github.com/spring-projects/spring-data-mongodb/blob/56e61a296582169c0498eda782889215500f9381/spring-data-mongodb/src/main/java/org/springframework/data/mongodb/repository/query/MongoQueryCreator.java